Add homekit_controller support for ecobee vendor extensions

josephnad opened the pull request:
homekit_controller now supports these vendor ecobee extensions:
  - Cancel Hold (button): resumes schedule on ecobee thermostat
  - Set Fan Mode (number): percentage where 0 means Auto, 100 means ON
  - Set Preset Mode (number): 0=Home, 1=Sleep, 2=Away

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #30258
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
homekit_controller support for ecobee vendor extensions (#30258)

homekit_controller now supports these vendor ecobee extensions:
  - Cancel Hold (button): resumes schedule on ecobee thermostat
  - Set Fan Mode (number): percentage where 0 means Auto, 100 means ON

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -46,17 +46,42 @@ class HomeKitButtonEntityDescription(ButtonEntityDescription):
     def async_add_characteristic(char: Characteristic):
-        if not (description := BUTTON_ENTITIES.get(char.type)):
-            return False
+        entities: list[HomeKitButton | HomeKitEcobeeButton] = []
```
You could just use a super class here (`ButtonEntity`). Although if the type signature on `async_setup_entry` was right (anywhere in homekit_controller) I don't think the annotation would be worth adding.

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -46,17 +46,42 @@ class HomeKitButtonEntityDescription(ButtonEntityDescription):
 }
 
 
+@dataclass
```
I wouldn't bother adding the skaffolding for generic ecobee buttons for one button type. As we are only expecting one button with this glitch. The declarative style is for when we are expecting lots of very similar characteristics with simple mappings to simple HA entity classes. But this is a workaround for a buggy characteristic, and we've only got one button with this particular bug. So making the declarative version is overkill (so far). It would be better to just do a "standard" entity class.

You define things like the device class directly on the class with the `_attr_` prefix. Then don't bother paramterizing `async_press` at all. Something like this:

```python
class EcobeeClearHoldButton(CharacteristicEntity, ButtonEntity):

    _attr_device_class = ButtonDeviceClass.UPDATE
    # ... snip

    def get_characteristic_types(self):
        """Define the homekit characteristics the entity is tracking."""
        return [CharacteristicsTypes.Vendor.ECOBEE_CLEAR_HOLD]

    # ... snip
    async def async_press(self) -> None:
         for val in (False, True):
            await self.async_put_characteristics({CharacteristicsTypes.Vendor.ECOBEE_CLEAR_HOLD: val})
```

You don't need to implement a description, and you don't need to pass it to this class. So you don't need to even subclass __init__.

Then have:

```python
BUTTON_ENTITY_CLASSES: dict[str, ButtonEntity] = {
    CharacteristicsTypes.Vendor.ECOBEE_CLEAR_HOLD: EcobeeClearHoldButton
}
```

We can make it declarative later if we are flooded with devices that recreate this bug.

Jc2k commented on homeassistant/components/homekit_controller/number.py:
```diff
@@ -88,3 +125,70 @@ async def async_set_value(self, value: float):
+
+    async def async_set_value(self, value: float):
+        """Set the characteristic to this value."""
+        if self.entity_description.first_send_different_value:
```
Same as with the button entity. I don't want to parameterize quirks too soon. Ecobee is the only device with these bugs. Make `ECOBEE_SET_HOLD_SCHEDULE` a normal number entity (as it doesn't seem to need quirks). Turn the remaining class into a class just for the fan speed.

Jc2k reviewed (changes requested):
This is a really good start. Thank you for looking into this. Your the first person both with a device and the drive to actually do something about it.

Is CharacteristicsTypes.Vendor.ECOBEE_SET_HOLD_SCHEDULE working as expected? It was previously reported that it does update the device, but home assistant sees the old state.

Jc2k commented:
A recent python update broke my dev environment, haven't had time to fix it yet, release will follow that.

josephnad commented:
> Is CharacteristicsTypes.Vendor.ECOBEE_SET_HOLD_SCHEDULE working as expected? It was previously reported that it does update the device, but home assistant sees the old state.

Setting preset mode to Home/Sleep/Away does not have the glitch and the thermostat switches to that preset mode from the first call. It was very consistent when I tested without any workarounds for this request, and no matter what the thermostat state was in.

I will go over your other comments and push another update with the changes.

josephnad commented:
@Jc2k, I pushed an update with your suggestions. It should resolve all your comments. I tested again to make sure the new code works without errors and as expected.

bdraco commented:
Looks like the CI didn't run correctly.  Will close/open to try to get it going

bdraco commented:
CI is running now.  This one needs to wait for Jc2k so I've removed the review request

Jc2k commented:
Hoping to get chance to sort out a release this afternoon, fingers crossed.

Jc2k commented:
0.6.5 is out. Once you have bumped the version in manifest.json, don't forget to run `./script/gen_requirements_all.py` to update all the places that reference the version.

josephnad commented:
> 0.6.5 is out. Once you have bumped the version in manifest.json, don't forget to run `./script/gen_requirements_all.py` to update all the places that reference the version.

Done. Pushed a commit with updated version of aiohomekit. Will keep a watch on CI results.

josephnad commented:
@Jc2k @bdraco , I think you need to approve my last change so that it runs the workflows.

Jc2k commented:
There you go

josephnad commented:
@Jc2k CI failed because I added a name() function to HomeKitNumber. Without the name(), the default name did not show the service name correctly in homeassistant. It was showing a generic name such as ecobee.number instead of ecobee.set_hold, where ecobee is the name of the thermostat. But this is causing tests for VOCOLINC to fail because it can not look up the name anymore.
Note that HomeKitButton was implemented with the name() function.
Was there any reason for HomeKitNumber to not override the name() function? Should I change the VOCOLINC test to use the new name, or just create a new class for the ecobee set hold entity?

Jc2k commented:
If the name of the vocolinc entity changed and it's a better name, fix the test.

josephnad commented:
I am trying to setup the test env locally, but the tox command fails to install async-timeout due to conflicting requirements:
    pyrmvtransport 0.3.3 depends on async-timeout>=4.0.0
    hangups 0.4.14 depends on async-timeout<4 and >=2

pyrmvtransport is set in homeassistant/components/rmvtransport/manifest.json
hangups version is set in homeassistant/components/hangouts/manifest.json

Jc2k commented:
TBH don't like tox, don't like it that much. I just install what I need to dev homekit. Something like:

```
pip install -e . -r requirements.txt -r requirements_test.txt
```

josephnad commented:
Does the github workflow also run test individually, or does it go through tox (installing all requirements)? I wanted to make sure it passes github workflow by testing things locally. I will try your approach with homekit only test.

Jc2k commented:
The GH version does things wildly different to maximise parallelisation.

josephnad commented:
ok i got it to run locally, and now I can see what the test is requesting number.vocolinc_flowerbud_0d324b, but with the added property name() in HomeKitNumber class, the entity is now called number.vocolinc_flowerbud_0d324b_spray_quantity.

This will be a breaking change for existing homekit numbers though. If a user has automation already using the older name.

But the HomeKitNumber implementation somehow assume only 1 instance for each vendor?

For ecobee, we have 2 number entities, and we need to have different names for the entities. So probably the best course of action is to just create another class for the ecobee set_hold number.

Jc2k commented:
I don't think it is a breaking change, because of the unique id? Because we use the entity registry (which is based on unique ids) our entities can be renamed. It's worth testing it in case i'm wrong.

josephnad commented:
I mean vocolinc name will change for this number entity, and this would be a breaking change for the users. 

When I go to Developer Tools->Services, I see the name of the number entity change each time I make a change to the name() function. So if we change the name() function to include the number name, vocolinc automation will break because now they would need to use the new name to set the value, i.e. instead of number.vocolinc_flowerbud_0d324b, they will need to specify number.vocolinc_flowerbud_0d324b_spray_quantity.

Jc2k commented:
So the name only sticks if you rename the entity? That's a shame. Still I think in this case I'd rather fix a bug (even if it's a breaking change) than have to keep adding new custom classes going forward.

josephnad commented:
somehow it preserves the name if it is changed later. so to test this properly, i had to remove the numbers, then delete them through the portal, then if I add them again without the name() override, HA is creating unique but vague names for those numbers such as "number.ecobee" and "number.ecobee_2".
For the vocolinc number, it doesn't include the name of the number, i.e. spray_quantity.
If you are okay to break this for the two numbers that are currently defined, I can just fix the test for vocolinc with the new name.

Jc2k commented:
They are both fairly obscure and fairly new. One is a config setting, not something you'd change with an automation. So go for it.

Jc2k commented:
This is the "somehow" you mentioned btw. I did a poor job of explaining it.

https://developers.home-assistant.io/docs/entity_registry_index/

> Being registered has the advantage that the same entity will always get the same entity ID. It will also prevent other entities from using that entity ID.

josephnad commented:
One cosmetic thing. 

In portal, if I go to Configuration->Devices and pick one of the ecobee devices, I can see the names of the button under Configuration, and the name of the numbers under Controls.
But, the names appear as "- Clear Hold" or "- Preset Mode". The "-" is coming from the name() function.

This is also same for the Sensors section "- Current Humidity" and "- Current Temperature".

In the dashboard, it shows the full name of the entity, so it is not an issue there.

Does it make sense to remove the "-" from the entity name()?

Jc2k commented:
For number, if it works better, go for it.

josephnad commented:
OK, I will keep the name with the "-" to be consistent with the rest.

I just pushed the change to fix the tests. Could you approve it so that the workflow start?

josephnad commented:
Looks like we are getting close, and only thing missing is adding test cases for code coverage to pass.

Jc2k commented:
Indeed. As well as `test_number` and `test_button`, it would be good to make sure the ecobee specific tests are updated for your new entities too. The tests in `specific_devices` are primarily enumeration tests. It's worth checking your device class etc is set correctly in the enumeration test (especially for the new classes).

josephnad commented:
@Jc2k This last commit failed on isort, but it is not clear to me why. isort runs when I commit the code locally, and it makes changes to the code, and then I can only commit when those changes are included. I am not sure if I need to resync my branch with the remote branch? If so, any pointers on what is the best way to pull code from remote? If I merge latest, it will go on top of my branch, which could cause problems to pull my request?

Jc2k commented:
I cloned your branch, did a fresh install with:

```
% git clone https://github.com/josephnad/core                           
Cloning into 'core'...
remote: Enumerating objects: 496784, done.
remote: Total 496784 (delta 0), reused 0 (delta 0), pack-reused 496784
Receiving objects: 100% (496784/496784), 278.64 MiB | 22.09 MiB/s, done.
Resolving deltas: 100% (353750/353750), done.
Updating files: 100% (21111/21111), done.
% cd core 
% python3 -m venv venv
% source venv/bin/activate
% python3 -m pip install -e . -r requirements.txt -r requirements_test.txt
Obtaining file:///private/tmp/core
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
SNIP SNIP SNIP
```

Then

```
% isort -rc homeassistant/components/homekit_controller/ tests/components/homekit_controller                                     
Fixing /private/tmp/core/homeassistant/components/homekit_controller/number.py
Fixing /private/tmp/core/homeassistant/components/homekit_controller/button.py
Fixing /private/tmp/core/tests/components/homekit_controller/specific_devices/test_ecobee3.py
/private/tmp/venv/lib/python3.9/site-packages/isort/main.py:1261: UserWarning: W0501: The following deprecated CLI flags were used and ignored: -rc!
  warn(
/private/tmp/venv/lib/python3.9/site-packages/isort/main.py:1265: UserWarning: W0500: Please see the 5.0.0 Upgrade guide: https://pycqa.github.io/isort/docs/upgrade_guides/5.0.0.html
  warn(
(venv) john@rangiku core % git diff
diff --git a/homeassistant/components/homekit_controller/button.py b/homeassistant/components/homekit_controller/button.py
index 3f647df802..54bd67fa08 100644
--- a/homeassistant/components/homekit_controller/button.py
+++ b/homeassistant/components/homekit_controller/button.py
@@ -9,6 +9,7 @@ from __future__ import annotations
 from dataclasses import dataclass
 
 from aiohomekit.model.characteristics import Characteristic, CharacteristicsTypes
+
 from homeassistant.components.button import (
     ButtonDeviceClass,
     ButtonEntity,
diff --git a/homeassistant/components/homekit_controller/number.py b/homeassistant/components/homekit_controller/number.py
index 24d9818f55..18cd47b263 100644
--- a/homeassistant/components/homekit_controller/number.py
+++ b/homeassistant/components/homekit_controller/number.py
@@ -7,6 +7,7 @@ characteristics that don't map to a Home Assistant feature.
 from __future__ import annotations
 
 from aiohomekit.model.characteristics import Characteristic, CharacteristicsTypes
+
 from homeassistant.components.number import NumberEntity, NumberEntityDescription
 from homeassistant.core import callback
 
diff --git a/tests/components/homekit_controller/specific_devices/test_ecobee3.py b/tests/components/homekit_controller/specific_devices/test_ecobee3.py
index 75de6522d0..3f7044a0d3 100644
--- a/tests/components/homekit_controller/specific_devices/test_ecobee3.py
+++ b/tests/components/homekit_controller/specific_devices/test_ecobee3.py
@@ -8,6 +8,7 @@ from unittest import mock
 
 from aiohomekit import AccessoryDisconnectedError
 from aiohomekit.testing import FakePairing
+
 from homeassistant.components.climate.const import (
     SUPPORT_TARGET_HUMIDITY,
     SUPPORT_TARGET_TEMPERATURE,
```

So I don't think you need to rebase to fix this, something else is going on. I'm not sure what. If it helps, I am running isort from the root of the checkout, and its reading its config from pyproject.toml.

josephnad commented:
That's weird. I activated a new env, and reinstalled python modules based on those requirements files, and the result of that isort command is touching a lot of files that I have not edited, and removing the empty line between imports. So clearly something else is happening here. I am using macos. I will try it on a linux box and see if I can get it to work as expected.

josephnad commented:
I was able to get isort to work properly on linux machine. Maybe something was messed up with my macos env. Anyway, the latest commit added the empty lines that workflow was checking for, and it should pass the check now, once you approve the workflow.

josephnad commented:
@jc2k All yours! :)

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -46,17 +46,67 @@ class HomeKitButtonEntityDescription(ButtonEntityDescription):
 }
 
 
+class HomeKitEcobeeClearHoldButton(CharacteristicEntity, ButtonEntity):
+    """Representation of a Button control for Ecobee clear hold request."""
+
+    def __init__(
+        self,
+        conn,
+        info,
+        char,
+    ):
+        """Initialise a HomeKit button control."""
+        self.entity_description = ButtonEntityDescription(
+            key=CharacteristicsTypes.Vendor.ECOBEE_CLEAR_HOLD,
+            name="Clear Hold",
+            device_class=ButtonDeviceClass.UPDATE,
+            entity_category=ENTITY_CATEGORY_CONFIG,
+        )
+        super().__init__(conn, info, char)
```
I _think_ we can hoist this out of __init__ and put make it a class variable and then completely get rid of the `__init__`.

```suggestion
class HomeKitEcobeeClearHoldButton(CharacteristicEntity, ButtonEntity):
    """Representation of a Button control for Ecobee clear hold request."""

    entity_description = ButtonEntityDescription(
        key=CharacteristicsTypes.Vendor.ECOBEE_CLEAR_HOLD,
        name="Clear Hold",
        device_class=ButtonDeviceClass.UPDATE,
        entity_category=ENTITY_CATEGORY_CONFIG,
    )
```

That said, i'm not 100% on the need to even use an entity description here. AIUI EntityDescriptions are when classes are data driven, and this one isn't. In this case the common pattern in other integrations is to do something like:

```suggestion
class HomeKitEcobeeClearHoldButton(CharacteristicEntity, ButtonEntity):
    """Representation of a Button control for Ecobee clear hold request."""
    
    _attr_name = "Clear Hold"
    _attr_device_class = ButtonDeviceClass.UPDATE
    _attr_entity_category = ENTITY_CATEGORY_CONFIG
```

(In this case you could just use `CharacteristicsTypes.Vendor.ECOBEE_CLEAR_HOLD` or `self._char.type` instead of `self.entity_description.key`).

If this works, can you make the same change to `HomeKitEcobeeFanModeNumber`.

josephnad replied:
I made those changes locally and this approach is working fine.

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -46,17 +46,67 @@ class HomeKitButtonEntityDescription(ButtonEntityDescription):
+
+    def get_characteristic_types(self):
+        """Define the homekit characteristics the entity is tracking."""
+        return [self.entity_description.key]
```
`get_characteristic_types` drives polling and even subscription, but this type is write only. Can you test returning an empty list here. This is probably the first time we've tried that so i'm not 100% it will work. It might be you need to include it in this list to be able to use it with put_char, but i've been trying to avoid that in recent refactorings IIRC). If it doesn't work, ignore this request and i'll come back to it in the future. Obv better to try it while there is someone with a device to test it!

```suggestion
        return []
```

Jc2k replied:
This change only makes sense for write only characteristics - for preset and fan mode we want the number to be readable as well as settable. So only do this for ClearHold.

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -46,17 +46,67 @@ class HomeKitButtonEntityDescription(ButtonEntityDescription):
+}
+
+
 async def async_setup_entry(hass, config_entry, async_add_entities):
```
It's a bit weird for `async_setup_entry` to be piggy in the middle like this. I think in a previous code review I was told to put `async_setup_entry` at the top, but it definitely shouldn't be in the middle.

josephnad replied:
I had an issue with the NUMBER_ENTITY_CLASSES when it mapped the char type to a class type, and the class was not yet defined. So I had to move these class definitions before this function. I can create another file to define these classes and import it?

josephnad replied:
Scratch that. I was able to move the mapping and classes to the bottom after async_setup_entry function. Works fine now.

Jc2k reviewed (changes requested):
Couple of nit picky things, one of which might not even work :) I think we are nearly there as far as I'm concerned but I need to find a solid bit of time to have a proper think over it. Will get to it as soon as I can..

josephnad commented:
@Jc2k does the latest commit look good? it should resolve all your requests.

Jc2k commented:
Looking good 👍 

Final things to check - are we happy with the entity category? The config category is meant for things like an Eve Degree's Elevation Above Sea Level setting (which is used to calibrate the pressure sensor). I wonder if for an Ecobee user this is a "primary entity" (something they'd expect to see on their main lovelace dashboard) (in which case we don't need to set the entity category).

Ditto for the device class. Looking at the docs for button (https://developers.home-assistant.io/docs/core/entity/button) i don't think there is a device class that makes sense for us, so maybe we should just not set it at all?

josephnad commented:
I removed the entity category and class for the clear hold button, and that's much better because it shows up in dashboard under  with the rest of entities.

josephnad commented:
@Jc2k looks like there are conflicts now with dev branch. I saw commits to button.py that needs to be resolved. I am not sure what is the right approach here. Should I force a push and collapse all these commits into a single commit on top of latest dev branch?
When do you plan to approve/merge this request?

Jc2k commented:
So it really depends on your reviewer. The general preference would be to go a rebase pull from `dev` but retain the commits, as it makes review easier. However, you'd likely have to keep fixing the same conflict. I'm personally fine with you squashing, then rebasing, then force pushing (we squash merge anyway).

I have the main repo as my `upstream` remote (`git remote add upstream git://github.com/home-assistant/core`) and i do `git pull --rebase upstream dev` (off the top of my head).

I'm hoping to get this merged in time for the next release, but there is no rush as there is no january release anyway. Things are super busy here, so I can't give a fixed timeline. I also plan to loop in a friend who helps QA the betas and has an Ecobee to take a look at this branch.

josephnad commented:
so if I merge on top of my existing commits, it creates a merge commit. Does git/github easily figure out what commits it needs to merge back to core from that? If so, I can keep the history and just resolve the conflicts.

So no release until February? hmm, I think I will need to find a way to install this branch on my raspberry pi so I can start using it, and migrating automation. 2 months is a long time :)

thecode commented:
Hi @josephnad,
Please rebase your branch to dev
Thanks

josephnad commented:
Hi @thecode,

The last update was rebased to dev. I rebased to master by mistake, but now this should be fixed.

Thanks

Jc2k commented on homeassistant/components/homekit_controller/number.py:
```diff
@@ -110,3 +90,98 @@ async def async_set_value(self, value: float):
+        icon="mdi:volume-high",
+        entity_category=EntityCategory.CONFIG,
+    ),
+    CharacteristicsTypes.Vendor.ECOBEE_SET_HOLD_SCHEDULE: NumberEntityDescription(
```
NumberEntity is meant for r/w entities, but according to the schema we have in the repo ECOBEE_SET_HOLD_SCHEDULE is w/o. Because of the limitation of these characteristics I think we need a custom entity that writes to `ECOBEE_SET_HOLD_SCHEDULE` but reads from `ECOBEE_CURRENT_MODE`.

I think if we are doing that it might be better to make a select entity (https://developers.home-assistant.io/docs/core/entity/select).

Maybe drop preset mode from this PR, and add it back in a new one? Then we can still hopefully get the others merged.

Jc2k replied:
PR: https://github.com/home-assistant/core/pull/65032

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -94,3 +83,51 @@ async def async_press(self) -> None:
+        """Return the name of the device if any."""
+        prefix = ""
+        if name := super().name:
+            prefix = f"{name} -"
```
This is fixed everywhere to not use `-` now, so drop the `-` here too.

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -94,3 +83,51 @@ async def async_press(self) -> None:
+            await self.async_put_characteristics({key: val})
+
+
+BUTTON_ENTITIES: dict[str, HomeKitButtonEntityDescription] = {
```
Can you leave this in its original place, even if that means its seperate from `BUTTON_ENTITY_CLASSES`. In other reviews i've been told to keep stuff like this (essentially a pile of constants) at the top, and I want to keep it consistent with sensor, switch, etc.

Jc2k commented on homeassistant/components/homekit_controller/number.py:
```diff
@@ -14,33 +14,6 @@
 
 from . import KNOWN_DEVICES, CharacteristicEntity
 
-NUMBER_ENTITIES: dict[str, NumberEntityDescription] = {
```
Can you leave this in its original place too.

Jc2k reviewed (changes requested):
Hey @josephnad - can you rebase when you've got a chance. I'd like to try and get this landed.

josephnad commented:
@Jc2k, I will take a look at this in next couple of days. I have a deliverable at work and will get the changes pushed on Saturday the latest.

josephnad commented:
@Jc2k, I rebased the code to upstream/dev, and squashed all commits to a single commit to simplify the code review.
You also mentioned that preset mode should be a separate commit and use a select entity, so I removed references to it.
Also moved the const back to where they were.
I will keep an eye on the integration test once the commit has been approved.

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -115,3 +123,45 @@ async def async_press(self) -> None:
+        prefix = ""
+        if name := super().name:
+            prefix = name
+        return f"{prefix} {self._attr_name}"
```
_attr_name is meant to be instead of implementing the name property - probably just put the string here directly given it's the only occurrence.

josephnad replied:
Done

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -115,3 +123,45 @@ async def async_press(self) -> None:
+# don't work!
+# We call get_uuid on *every* type to normalise them to the long form
+# Eventually aiohomekit will use the long form exclusively amd this can be removed.
+for k, v in list(BUTTON_ENTITY_CLASSES.items()):
```
Don't bother with this cludge. Its not needed here because clear hold is already a normalised uuid. And i'm working to get rid of all the other occurrences of this this next release cycle so there is no point in adding it.

josephnad replied:
Removed

Jc2k commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -115,3 +123,45 @@ async def async_press(self) -> None:
+        """Press the button."""
+        key = self._char.type
+
+        for val in (False, True):
```
Maybe add a comment that this is (we think?) to clear incorrectly cached state in the ecobee, so the 2nd  put actually takes?

josephnad replied:
added comment

Jc2k commented on homeassistant/components/homekit_controller/number.py:
```diff
@@ -152,3 +159,66 @@ async def async_set_value(self, value: float):
+        else:
+            other_value = self.min_value + 1
+
+        if value != other_value:
```
Ditto for a comment about what we are doing here.

josephnad replied:
added comment

Jc2k commented on homeassistant/components/homekit_controller/number.py:
```diff
@@ -152,3 +159,66 @@ async def async_set_value(self, value: float):
+        prefix = ""
+        if name := super().name:
+            prefix = name
+        return f"{prefix} {self._attr_name}"
```
See my other _attr_name comment.

josephnad replied:
removed _attr_name

Jc2k reviewed (changes requested):
Couple of tiny things that would be good to get in then I am happy.

(I'm about to overhaul the tests for the entire component so I haven't been too thorough with review there).

Jc2k reviewed (approved):
LGTM

bdraco commented on homeassistant/components/homekit_controller/button.py:
```diff
@@ -115,3 +123,37 @@ async def async_press(self) -> None:
+class HomeKitEcobeeClearHoldButton(CharacteristicEntity, ButtonEntity):
+    """Representation of a Button control for Ecobee clear hold request."""
+
+    def get_characteristic_types(self):
```
Missing return typing

Jc2k replied:
The return type of this is about to change, FWIW.

Jc2k replied:
(To elaborate, when this PR is merged i'll be rebasing 3+ branches which move us towards CharacteristicsTypes and ServicesTypes being proper enums so we can use them in types here, at which point i'll be updating all the integration and tests to use them. So i'm a +0 about typing homekit_controller internals right now.)

bdraco commented on homeassistant/components/homekit_controller/number.py:
```diff
@@ -152,3 +159,72 @@ async def async_set_value(self, value: float):
+class HomeKitEcobeeFanModeNumber(CharacteristicEntity, NumberEntity):
+    """Representation of a Number control for Ecobee Fan Mode request."""
+
+    def get_characteristic_types(self):
```
Missing return typing

bdraco commented on homeassistant/components/homekit_controller/number.py:
```diff
@@ -152,3 +159,72 @@ async def async_set_value(self, value: float):
+        """Return the current characteristic value."""
+        return self._char.value
+
+    async def async_set_value(self, value: float):
```
Missing return typing
